hoomd/mpcd/SphereGeometry.h

```diff
@@ -0,0 +1,209 @@
+         *       amount of time remaining.
+         */
+        HOSTDEVICE bool detectCollision(Scalar3& pos, Scalar3& vel, Scalar& dt) const
+            {
+
+            /*
+             * Detect if particle has left the spherical confinement.
+             * comparison: 'r2 < R*R' used in the following calculations is equal to 0(FALSE) if the particle is outside
+             * a sphere of radius R, 1(TRUE) if inside.
+             *
+             * We intentionally use < rather than <= in the comparison so that spurious collisions are not detected
+             * when a particle is reset to the spherical boundary. A particle landing exactly on the spherical boundary
+             * can be reflected in the next streaming step.
+             * TODO: Verify if the comparison can impact the dynamics incorrectly since we aim to have a dynamic spherical confinement.
+             */
+            Scalar r2 = dot(pos,pos);
+            // exit immediately if no collision is found or particle is not moving normal (radial direction) to the wall
+            // (since no new collision could have occurred if there is no normal motion)
+            /*
+            r(t) = sqrt(x**2 + y**2 + z**2)
+            radial velocity, r_dot = (x*vel_x + y*vel_y + z*vel_z)/r(t)
+
+            define variable vr = dot(pos,vel) to check for radial motion, for computational efficiency.
+
+            To avoid division by zero error later on (line 103), we exit immediately. (no collision could have occurred
+            if the particle speed is equal to zero)
+            */
+            Scalar vr = dot(pos,vel);
+            Scalar v2 = dot(vel,vel);
+            if (r2 < R*R || vr == Scalar(0) || v2 == Scalar(0))
+               {
+               dt = Scalar(0);
+               return false;
+               }
+
+            // Find the point of contact when the particle is just leaving the surface
+            /*
+            * Calculating the time spent outside-bounds requires the knowledge of the point of contact on the boundary.
+            * The point of contact can be calculated using geometrical considerations.
+            *
+            * We know the following quantities,
+            *    1. r(t+del_t)      (the point outside the sphere)
+            *    2. v(t)            (previous velocity)
+            *
+            * Assuming 'r*' is the point of contact on the spherical shell and dx* is the distance travelled by the
+            * particle outside the boundary.
+            *
+            * r* = r(t+del_t) - dx* * v(t)/|v|
+            * and |r|**2 = R**2
+            *
+            * Solving the above equations,
+            * dx* = (r.v^) - sqrt((r.v^)**2 - r2 + R2)
+            * consequently, dt = dx* / |v|
+            */
+
+            Scalar mod_v = fast::sqrt(v2);
+            Scalar r_dot_v_hat = vr/mod_v;
+
+            const Scalar dx_ = r_dot_v_hat - fast::sqrt(r_dot_v_hat*r_dot_v_hat - r2 + R*R);
+            dt = dx_/mod_v;
```

Comment:
Using the same math you did, you can solve for `dt` directly without needing to get the norm of the velocity. Write `r* = r - v dt` and solve for `dt` (or manipulate the equations you solved to go from dx to dt). I think the right result is:
```
const Scalar rv = dot(r,v);
dt = (rv - fast::sqrt(rv*rv-v2*(r2-R*R)))/v2;
```
Please check if this is right. You should also move the `dot(r,v)` to here rather than above, since it is not needed until this point.

Reply:
I have checked your calculation, and it should be correct.